frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -60,33 +73,134 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) -> Result<(), MinerError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	match storage.mutate::<_, (), _>(|_| Ok(call.clone())) {
+		Ok(Ok(_)) => Ok(()),
+		Ok(Err(_)) => Err(MinerError::FailedToStoreSolution),
+		Err(_) => unreachable!("mutation function cannot fail; qed"),
+	}
+}
+
+/// Get a saved solution from OCW storage if it exists.
+fn restore_solution<T: Config>() -> Result<Call<T>, MinerError> {
+	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL)
+		.get()
+		.flatten()
+		.ok_or(MinerError::NoStoredSolution)
+}
+
+/// Clear a saved solution from OCW storage.
+pub(super) fn kill_solution<T: Config>() {
+	let mut storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	storage.clear();
+}
+
 impl<T: Config> Pallet<T> {
-	/// Mine a new solution, and submit it back to the chain as an unsigned transaction.
-	pub fn mine_check_and_submit() -> Result<(), MinerError> {
+	/// Attempt to restore a solution from cache. Otherwise, compute it fresh. Either way, submit
+	/// if our call's score is greater than that of the cached solution.
+	pub fn restore_or_compute_then_maybe_submit() -> Result<(), MinerError> {
+		log!(
+			info,
+			"OCW attempting to restore or compute an unsigned solution for the current election"
+		);
+
+		let (call, score) = restore_solution::<T>()
+			.and_then(|call| {
+				// ensure the cached call is still current before submitting
+				let current_round = Self::round();
+				match &call {
+					Call::submit_unsigned(solution, _) if solution.round == current_round => {
+						let score = solution.score.clone();
+
+						log!(
+							debug,
+							"restored a cached call for round {} with score {:?}",
+							current_round,
+							score,
+						);
+
+						Ok((call, score))
+					}
+					_ => Err(MinerError::SolutionOutOfDate),
+				}
+			})
+			.or_else::<MinerError, _>(|_| {
+				// if not present or cache invalidated, regenerate
+				let (call, score) = Self::mine_checked_call()?;
+				save_solution(&call)?;
+				Ok((call, score))
+			})?;
+
+		if Self::queued_solution().map_or(true, |q: ReadySolution<_>| is_score_better::<Perbill>(
+			score,
+			q.score,
+			T::SolutionImprovementThreshold::get()
+		)) {
+			Self::submit_call(call)
+		} else {
+			log!(
+				info,
```

Comment:
I think you should revise your log levels: info logs will be shown to the end user and must be only things that matter to them. All of these are more like `debug` to me.

Reply:
https://github.com/paritytech/substrate/pull/8290/commits/5eb2673a7a2b06806a44f87c09080192b5687cd3